File: backend/src/main/java/com/improvementsolutions/service/inventory/CambioEppSolicitudService.java
package com.improvementsolutions.service.inventory;

import java.time.LocalDateTime;
import java.time.Year;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.improvementsolutions.model.Business;
import com.improvementsolutions.model.inventory.CambioEppSolicitud;
import com.improvementsolutions.repository.inventory.CambioEppSolicitudRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CambioEppSolicitudService {

    private static final List<String> PENDIENTES_ENTREGA = List.of("EN_ESPERA_INVENTARIO", "VALIDADO_FIRMADO");

    private final CambioEppSolicitudRepository repository;
    private final InventoryAuthorizationService authService;
    private final ObjectMapper objectMapper;

    @Transactional(readOnly = true)
    public List<Map<String, Object>> list(String ruc) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        return repository.findByBusiness_IdOrderByCreatedAtDesc(business.getId())
                .stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listPendientesEntrega(String ruc) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        return repository.findByBusiness_IdAndStatusInOrderByCreatedAtDesc(business.getId(), PENDIENTES_ENTREGA)
                .stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    /** Tipos de acontecimiento asignados a la empresa (Inventario-Bodega). */
    @Transactional(readOnly = true)
    public List<Map<String, Object>> listAcontecimientoTypesForBusiness(String ruc) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        if (business.getInventoryAcontecimientoTypes() != null) {
            business.getInventoryAcontecimientoTypes().size();
        }
        List<Map<String, Object>> out = new java.util.ArrayList<>();
        if (business.getInventoryAcontecimientoTypes() == null) {
            return out;
        }
        business.getInventoryAcontecimientoTypes().stream()
                .filter(t -> t != null && !Boolean.FALSE.equals(t.getActive()))
                .sorted((a, b) -> String.valueOf(a.getName()).compareToIgnoreCase(String.valueOf(b.getName())))
                .forEach(t -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("id", t.getId());
                    m.put("name", t.getName());
                    m.put("description", t.getDescription());
                    out.add(m);
                });
        return out;
    }

    /** Estados del EPI asignados a la empresa (Inventario-Bodega). */
    @Transactional(readOnly = true)
    public List<Map<String, Object>> listEstadoEpiForBusiness(String ruc) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        if (business.getInventoryEstadoEpis() != null) {
            business.getInventoryEstadoEpis().size();
        }
        List<Map<String, Object>> out = new java.util.ArrayList<>();
        if (business.getInventoryEstadoEpis() == null) {
            return out;
        }
        business.getInventoryEstadoEpis().stream()
                .filter(t -> t != null && !Boolean.FALSE.equals(t.getActive()))
                .sorted((a, b) -> String.valueOf(a.getName()).compareToIgnoreCase(String.valueOf(b.getName())))
                .forEach(t -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("id", t.getId());
                    m.put("name", t.getName());
                    m.put("description", t.getDescription());
                    out.add(m);
                });
        return out;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getById(String ruc, Long id) {
        return toDto(requireOwned(ruc, id));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> nextReportNumber(String ruc) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        int year = Year.now().getValue();
        String prefix = "REP-" + year + "-";
        long count = repository.countByBusinessAndReportePrefix(business.getId(), prefix + "%");
        String nReporte = String.format("%s%05d", prefix, count + 1);
        // Evitar colisión si hubo huecos / borrados
        int guard = 0;
        while (repository.existsByBusinessIdAndReportNumber(business.getId(), nReporte) && guard < 1000) {
            count++;
            nReporte = String.format("%s%05d", prefix, count + 1);
            guard++;
        }
        return Map.of("nReporte", nReporte, "year", year);
    }

    @Transactional
    public Map<String, Object> create(String ruc, Map<String, Object> body) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        CambioEppSolicitud entity = new CambioEppSolicitud();
        entity.setBusiness(business);
        applyCreateFields(entity, body);

        if (entity.getNReporte() == null || entity.getNReporte().isBlank()) {
            entity.setNReporte(String.valueOf(nextReportNumber(ruc).get("nReporte")));
        }
        if (repository.existsByBusinessIdAndReportNumber(business.getId(), entity.getNReporte())) {
            throw new IllegalArgumentException("Ya existe una solicitud con el número " + entity.getNReporte());
        }
        entity.setStatus(normalizeStatus(asString(body.get("status"), "PENDIENTE_APROBACION")));
        return toDto(repository.save(entity));
    }

    @Transactional
    public Map<String, Object> markEnEsperaInventario(String ruc, Long id, Map<String, Object> body) {
        CambioEppSolicitud entity = requireOwned(ruc, id);
        String path = asString(body.get("signedFilePath"), "").trim();
        String name = asString(body.get("signedFileName"), "").trim();
        if (path.isEmpty()) {
            throw new IllegalArgumentException("Falta la ruta del documento firmado (signedFilePath).");
        }
        entity.setSignedFilePath(path);
        entity.setSignedFileName(name.isEmpty() ? path.substring(path.lastIndexOf('/') + 1) : name);
        entity.setSignedUploadedAt(LocalDateTime.now());
        entity.setStatus("EN_ESPERA_INVENTARIO");
        entity.setRejectedReason(null);
        entity.setRejectedAt(null);
        return toDto(repository.save(entity));
    }

    @Transactional
    public Map<String, Object> markEntregado(String ruc, Long id, Map<String, Object> body) {
        CambioEppSolicitud entity = requireOwned(ruc, id);
        if (!"EN_ESPERA_INVENTARIO".equals(entity.getStatus()) && !"VALIDADO_FIRMADO".equals(entity.getStatus())) {
            throw new IllegalStateException("La solicitud no está en espera de inventario (estado: " + entity.getStatus() + ").");
        }
        Long outputId = null;
        Object rawOut = body.get("outputId");
        if (rawOut instanceof Number n) outputId = n.longValue();
        else if (rawOut != null && !String.valueOf(rawOut).isBlank()) {
            try { outputId = Long.parseLong(String.valueOf(rawOut)); } catch (Exception ignored) {}
        }
        entity.setOutputId(outputId);
        entity.setOutputNumber(asString(body.get("outputNumber"), null));
        entity.setDeliveredAt(LocalDateTime.now());
        entity.setStatus("VALIDADO_ENTREGADO");
        return toDto(repository.save(entity));
    }

    @Transactional
    public Map<String, Object> markRechazado(String ruc, Long id, Map<String, Object> body) {
        CambioEppSolicitud entity = requireOwned(ruc, id);
        String motivo = asString(body.get("rejectedReason"), asString(body.get("reason"), "")).trim();
        if (motivo.length() < 5) {
            throw new IllegalArgumentException("Indique el motivo del rechazo (mínimo 5 caracteres).");
        }
        entity.setRejectedReason(motivo);
        entity.setRejectedAt(LocalDateTime.now());
        entity.setStatus("RECHAZADO");
        return toDto(repository.save(entity));
    }

    private CambioEppSolicitud requireOwned(String ruc, Long id) {
        Business business = authService.requireBusinessForRucAndCurrentUser(ruc);
        return repository.findByBusiness_IdAndId(business.getId(), id)
                .orElseThrow(() -> new IllegalArgumentException("Solicitud de cambio EPP no encontrada."));
    }

    private void applyCreateFields(CambioEppSolicitud entity, Map<String, Object> body) {
        entity.setNReporte(asString(body.get("nReporte"), null));
        entity.setTrabajador(asString(body.get("trabajador"), "").trim());
        entity.setCedula(asString(body.get("cedula"), "").trim());
        if (entity.getTrabajador().isEmpty() || entity.getCedula().isEmpty()) {
            throw new IllegalArgumentException("Trabajador y cédula son obligatorios.");
        }
        entity.setCargo(asString(body.get("cargo"), null));
        entity.setArea(asString(body.get("area"), null));
        entity.setTipoAcontecimiento(asString(body.get("tipoAcontecimiento"), null));
        entity.setSectionCode(asString(body.get("sectionCode"), null));
        entity.setSectionLabel(asString(body.get("sectionLabel"), null));
        entity.setEstadoEpi(asString(body.get("estadoEpi"), null));
        entity.setSeveridad(asString(body.get("severidad"), null));
        entity.setFechaElaboracion(asString(body.get("fechaElaboracion"), null));

        Object snap = body.get("formSnapshot");
        if (snap != null) {
            try {
                entity.setFormSnapshot(snap instanceof String s ? s : objectMapper.writeValueAsString(snap));
            } catch (Exception e) {
                throw new IllegalArgumentException("formSnapshot inválido.");
            }
        }
    }

    private String normalizeStatus(String status) {
        String s = status == null ? "PENDIENTE_APROBACION" : status.trim();
        if ("VALIDADO_FIRMADO".equals(s)) return "EN_ESPERA_INVENTARIO";
        if ("ENTREGADO".equals(s)) return "VALIDADO_ENTREGADO";
        return s;
    }

    private Map<String, Object> toDto(CambioEppSolicitud e) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", e.getId() == null ? null : String.valueOf(e.getId()));
        m.put("nReporte", e.getNReporte());
        m.put("trabajador", e.getTrabajador());
        m.put("cedula", e.getCedula());
        m.put("cargo", e.getCargo());
        m.put("area", e.getArea());
        m.put("tipoAcontecimiento", e.getTipoAcontecimiento());
        m.put("sectionCode", e.getSectionCode());
        m.put("sectionLabel", e.getSectionLabel());
        m.put("estadoEpi", e.getEstadoEpi());
        m.put("severidad", e.getSeveridad());
        m.put("fechaElaboracion", e.getFechaElaboracion());
        m.put("status", normalizeStatus(e.getStatus()));
        m.put("formSnapshot", parseSnapshot(e.getFormSnapshot()));
        m.put("signedFileName", e.getSignedFileName());
        m.put("signedFilePath", e.getSignedFilePath());
        m.put("signedUploadedAt", e.getSignedUploadedAt() == null ? null : e.getSignedUploadedAt().toString());
        m.put("outputId", e.getOutputId());
        m.put("outputNumber", e.getOutputNumber());
        m.put("deliveredAt", e.getDeliveredAt() == null ? null : e.getDeliveredAt().toString());
        m.put("rejectedReason", e.getRejectedReason());
        m.put("rejectedAt", e.getRejectedAt() == null ? null : e.getRejectedAt().toString());
        m.put("createdAt", e.getCreatedAt() == null ? null : e.getCreatedAt().toString());
        m.put("updatedAt", e.getUpdatedAt() == null ? null : e.getUpdatedAt().toString());
        return m;
    }

    private Map<String, Object> parseSnapshot(String raw) {
        if (raw == null || raw.isBlank()) return Map.of();
        try {
            return objectMapper.readValue(raw, new TypeReference<Map<String, Object>>() {});
        } catch (Exception e) {
            return Map.of();
        }
    }

    private static String asString(Object v, String def) {
        if (v == null) return def;
        String s = String.valueOf(v).trim();
        return s.isEmpty() || "null".equalsIgnoreCase(s) ? def : s;
    }
}

File: frontend/src/app/services/cambio-epp-solicitud.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of, throwError } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { FileService } from './file.service';

export interface CambioEppFormSnapshot {
  empresa: string;
  tipoAcontecimiento: string;
  sectionCode: string;
  sectionLabel: string;
  nReporte: string;
  nombreTrabajador: string;
  cedula: string;
  cargo: string;
  area: string;
  estadoEpi: string;
  fechaElaboracion: string;
  fechaNotificacion: string;
  fechaDotacion: string;
  talla: string;
  accionSuspension: boolean;
  accionMulta: boolean;
  accionNotificacion: boolean;
  accionCambio: boolean;
  severidad: string;
  observaciones: string;
  nombreEmisor: string;
  cargoEmisor: string;
}

/**
 * Flujo persistido en BD (API /api/inventory/{ruc}/cambio-epp):
 * 1) PENDIENTE_APROBACION — PDF generado; falta subir firmado
 * 2) EN_ESPERA_INVENTARIO — firmado subido; Inventario revisa en Nueva Salida
 * 3) VALIDADO_ENTREGADO — Inventario validó y doto el EPP
 * 4) RECHAZADO — Inventario canceló/rechazó con motivo
 */
export type CambioEppSolicitudStatus =
  | 'PENDIENTE_APROBACION'
  | 'EN_ESPERA_INVENTARIO'
  | 'VALIDADO_FIRMADO'
  | 'VALIDADO_ENTREGADO'
  | 'ENTREGADO'
  | 'RECHAZADO';

export interface CambioEppSolicitud {
  id: string;
  nReporte: string;
  trabajador: string;
  cedula: string;
  cargo: string;
  area: string;
  tipoAcontecimiento: string;
  sectionCode?: string;
  sectionLabel?: string;
  estadoEpi: string;
  severidad: string;
  fechaElaboracion: string;
  status: CambioEppSolicitudStatus;
  createdAt: string;
  formSnapshot: CambioEppFormSnapshot;
  signedFileName?: string;
  signedFilePath?: string;
  signedUploadedAt?: string;
  outputId?: number;
  outputNumber?: string;
  deliveredAt?: string;
  rejectedReason?: string;
  rejectedAt?: string;
}

export interface CambioEppSignedFile {
  name: string;
  type: string;
  dataUrl?: string;
  blob?: Blob;
  path?: string;
}

@Injectable({ providedIn: 'root' })
export class CambioEppSolicitudService {
  constructor(
    private http: HttpClient,
    private fileService: FileService
  ) {}

  private base(ruc: string): string {
    return `/api/inventory/${ruc}/cambio-epp`;
  }

  list(ruc: string): Observable<CambioEppSolicitud[]> {
    return this.http.get<any[]>(this.base(ruc)).pipe(
      map(rows => (rows || []).map(x => this.normalize(x))),
      catchError(() => of([]))
    );
  }

  listPendientesEntrega(ruc: string): Observable<CambioEppSolicitud[]> {
    return this.http.get<any[]>(`${this.base(ruc)}/pendientes-entrega`).pipe(
      map(rows => (rows || []).map(x => this.normalize(x))),
      catchError(() => of([]))
    );
  }

  /** Tipos de acontecimiento asignados a la empresa (Inventario-Bodega). */
  listTiposAcontecimiento(ruc: string): Observable<Array<{ id?: number; name: string; description?: string }>> {
    return this.http.get<Array<{ id?: number; name: string; description?: string }>>(
      `${this.base(ruc)}/tipos-acontecimiento`
    ).pipe(catchError(() => of([])));
  }

  /** Estados del EPI asignados a la empresa (Inventario-Bodega). */
  listEstadosEpi(ruc: string): Observable<Array<{ id?: number; name: string; description?: string }>> {
    return this.http.get<Array<{ id?: number; name: string; description?: string }>>(
      `${this.base(ruc)}/estados-epi`
    ).pipe(catchError(() => of([])));
  }

  nextReportNumber(ruc: string): Observable<string> {
    return this.http.get<{ nReporte?: string }>(`${this.base(ruc)}/next-report-number`).pipe(
      map(r => r?.nReporte || `REP-${new Date().getFullYear()}-00001`),
      catchError(() => of(`REP-${new Date().getFullYear()}-00001`))
    );
  }

  create(ruc: string, item: Partial<CambioEppSolicitud>): Observable<CambioEppSolicitud> {
    return this.http.post<any>(this.base(ruc), {
      nReporte: item.nReporte,
      trabajador: item.trabajador,
      cedula: item.cedula,
      cargo: item.cargo,
      area: item.area,
      tipoAcontecimiento: item.tipoAcontecimiento,
      sectionCode: item.sectionCode,
      sectionLabel: item.sectionLabel,
      estadoEpi: item.estadoEpi,
      severidad: item.severidad,
      fechaElaboracion: item.fechaElaboracion,
      status: item.status || 'PENDIENTE_APROBACION',
      formSnapshot: item.formSnapshot || this.emptySnapshot()
    }).pipe(map(x => this.normalize(x)));
  }

  /** Sube PDF firmado a disco y marca EN_ESPERA_INVENTARIO en BD. */
  uploadSignedAndMarkEnEspera(ruc: string, id: string, file: File): Observable<CambioEppSolicitud> {
    return this.fileService.uploadFileToDirectory('cambio_epp', file).pipe(
      switchMap(resp => {
        const path = resp?.url || '';
        if (!path) {
          return throwError(() => new Error('No se pudo subir el documento firmado.'));
        }
        return this.http.patch<any>(`${this.base(ruc)}/${id}/en-espera-inventario`, {
          signedFilePath: path,
          signedFileName: file.name || resp.filename || 'firmado.pdf'
        });
      }),
      map(x => this.normalize(x))
    );
  }

  markEntregado(ruc: string, id: string, outputId?: number, outputNumber?: string): Observable<CambioEppSolicitud> {
    return this.http.patch<any>(`${this.base(ruc)}/${id}/entregado`, {
      outputId: outputId ?? null,
      outputNumber: outputNumber || null
    }).pipe(map(x => this.normalize(x)));
  }

  markRechazado(ruc: string, id: string, reason: string): Observable<CambioEppSolicitud> {
    return this.http.patch<any>(`${this.base(ruc)}/${id}/rechazar`, {
      rejectedReason: reason
    }).pipe(map(x => this.normalize(x)));
  }

  /** Obtiene el PDF firmado desde el servidor (para ver / reenviar a salida). */
  getSignedFile(ruc: string, solicitud: CambioEppSolicitud): Observable<CambioEppSignedFile | null> {
    const path = String(solicitud?.signedFilePath || '').trim();
    if (!path) return of(null);
    const url = this.fileService.getFileUrl(path);
    return this.fileService.downloadFile(url).pipe(
      map(blob => ({
        name: solicitud.signedFileName || 'documento-firmado.pdf',
        type: blob.type || 'application/pdf',
        blob,
        path
      })),
      catchError(() => of(null))
    );
  }

  signedFileToFile(signed: CambioEppSignedFile, fallbackName: string): File | null {
    if (!signed?.blob) return null;
    return new File(
      [signed.blob],
      signed.name || fallbackName || 'firmado.pdf',
      { type: signed.type || 'application/pdf' }
    );
  }

  /** Convierte blob a object URL para visor (revocar después). */
  blobToObjectUrl(blob: Blob): string {
    return URL.createObjectURL(blob);
  }

  isEnEsperaInventario(status: string | undefined): boolean {
    return status === 'EN_ESPERA_INVENTARIO' || status === 'VALIDADO_FIRMADO';
  }

  isValidadoEntregado(status: string | undefined): boolean {
    return status === 'VALIDADO_ENTREGADO' || status === 'ENTREGADO';
  }

  emptySnapshot(): CambioEppFormSnapshot {
    return {
      empresa: '',
      tipoAcontecimiento: '',
      sectionCode: '',
      sectionLabel: '',
      nReporte: '',
      nombreTrabajador: '',
      cedula: '',
      cargo: '',
      area: '',
      estadoEpi: '',
      fechaElaboracion: '',
      fechaNotificacion: '',
      fechaDotacion: '',
      talla: '',
      accionSuspension: false,
      accionMulta: false,
      accionNotificacion: false,
      accionCambio: true,
      severidad: '',
      observaciones: '',
      nombreEmisor: '',
      cargoEmisor: ''
    };
  }

  private normalize(x: any): CambioEppSolicitud {
    const snap = x?.formSnapshot && typeof x.formSnapshot === 'object'
      ? { ...this.emptySnapshot(), ...x.formSnapshot }
      : this.emptySnapshot();
    let status = String(x?.status || 'PENDIENTE_APROBACION');
    if (status === 'VALIDADO_FIRMADO') status = 'EN_ESPERA_INVENTARIO';
    if (status === 'ENTREGADO') status = 'VALIDADO_ENTREGADO';
    return {
      id: String(x?.id ?? ''),
      nReporte: x?.nReporte || '',
      trabajador: x?.trabajador || '',
      cedula: x?.cedula || '',
      cargo: x?.cargo || '',
      area: x?.area || '',
      tipoAcontecimiento: x?.tipoAcontecimiento || '',
      sectionCode: x?.sectionCode || '',
      sectionLabel: x?.sectionLabel || '',
      estadoEpi: x?.estadoEpi || '',
      severidad: x?.severidad || '',
      fechaElaboracion: x?.fechaElaboracion || '',
      status: status as CambioEppSolicitudStatus,
      createdAt: x?.createdAt || '',
      formSnapshot: snap,
      signedFileName: x?.signedFileName || undefined,
      signedFilePath: x?.signedFilePath || undefined,
      signedUploadedAt: x?.signedUploadedAt || undefined,
      outputId: x?.outputId ?? undefined,
      outputNumber: x?.outputNumber || undefined,
      deliveredAt: x?.deliveredAt || undefined,
      rejectedReason: x?.rejectedReason || undefined,
      rejectedAt: x?.rejectedAt || undefined
    };
  }
}
